### PR TITLE
[pruning][core][feature] LSTM Structured Pruning prune_functions + pattern

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90801

Summary:

This PR adds in support for LSTM Structured Pruning. 

- Adds in LSTMSaliencyPruner, an implemented pruner that splits the packed weights, finds the appropriate mask for each piece individually based on saliency, and then combines to create an overall mask for the LSTM.  
- Adds in pruning functions for LSTM pruning, which will split the weights, apply the masks, and then recombine the pruned weights. Works for both single and multiple-layer LSTMs.

Also added a basic pattern to the default set of of patterns for
LSTM -> Linear pruning
LSTM -> LayerNorm -> Linear pruning

Adds in test to check that LSTM pruning works, as well as for LSTMSaliencyPruner

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_single_layer`
`python test/test_ao_sparsity.py -- TestBaseStructuredSparsifier.test_prune_lstm_layernorm_linear_multiple_layer`
`python test/test_ao_sparsity.py -- TestSaliencyPruner.test_lstm_saliency_pruner_update_mask`
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42199001](https://our.internmc.facebook.com/intern/diff/D42199001)

### CODE REVIEW

File: test/ao/sparsity/test_structured_sparsifier.py [hunk 1]
```diff
@@ -639,3 +663,92 @@ def test_complex_conv2d(self):
...
+        fx_pruner.step()
+
+        model.eval()
+        _, _ = model(rnn_input)
+        pruned_model = fx_pruner.prune()
+        pruned_model.eval()
+        _, _ = pruned_model(rnn_input)
+
+        expected_params = dict(model.named_parameters())
+        for name, param in model.named_parameters():
```
Comment (jcaip) [inline test/ao/sparsity/test_structured_sparsifier.py:699]: If anyone has a better idea about how to test this, please let me know.

Comment (WongJohnson) [reply test/ao/sparsity/test_structured_sparsifier.py:699]: It appears you are testing the parameters against themselves there? Do you mean to iterate over the pruned_model?

File: torch/testing/_internal/common_pruning.py [hunk 2]
```diff
@@ -7,6 +7,10 @@
 from torch import nn
 
 
+def elements_are_subset(subset_tensor, superset_tensor):
```
Comment (WongJohnson) [inline torch/testing/_internal/common_pruning.py:10]: I didn't follow the rest of your code, but I assume pruning here means that certain "rows" are simply removed from the superset_tensor. If that's the case, you can search for the `subset_tensor` as you sequentially sweep through `superset_tensor` to avoid a lot of duplicate comparisons (and also avoid some edge case false True returns, i.e. matching the same row multiple times):

```
def elements_are_subset_ordered(subset_tensor, superset_tensor):
    i = 0
    for row in subset_tensor:  # or iterate over whichever dimension is appropriate
        while i < len(superset_tensor):
            if not torch.equal(row, superset_tensor[i]):
                i += 1  # keep searching
            else:
                break  # check next row (continue for loop)
        else:
            return False  # searched beyond end of superset_tensor

    return True  # matched all in subset_tensor
```

I'm happy to discuss if I misunderstood the assumptions here.

File: test/ao/sparsity/test_structured_sparsifier.py (hunk 1 above)
Comment (WongJohnson) [reply test/ao/sparsity/test_structured_sparsifier.py:699]: If so, I believe this is equivalent to your test:
```
pruned_params = dict(pruned_model.named_parameters())
for name, param in model.named_parameters():
    assert name in pruned_params, f"pruned param {name} is missing"
    assert elements_are_subset(pruned_params[name], param), f"pruned param {name} ({pruned_params[name].shape}) is not a subset of model ({param.shape})"
```

Comment (jcaip) [reply test/ao/sparsity/test_structured_sparsifier.py:699]: thanks, good catch!

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

File: test/ao/sparsity/test_structured_sparsifier.py [hunk 3]
```diff
@@ -75,6 +95,72 @@ def test_update_mask(self):
         assert expected.shape == pruned.shape
         assert torch.isclose(expected, pruned, rtol=1e-05, atol=1e-07).all()
 
+    def test_lstm_saliency_pruner_update_mask(self):
```
Comment (jcaip) [inline test/ao/sparsity/test_structured_sparsifier.py:98]: @WongJohnson I added this test for the SaliencyPruner and  two additional tests for LSTM -> layernorm -> Linear below.

Comment (WongJohnson) [reply test/ao/sparsity/test_structured_sparsifier.py:98]: Great, thanks for adding this. I'd like to see a bit more complex tests (perhaps randomly-drawn tensors) but I'll leave that up to your team's judgement. I would at least ensure that you test properly handling negative weights and also 1d tensors (i.e. bias tensors for which I had to add the dims() check - I believe `SaliencyPruner` still has this issue).

File: torch/testing/_internal/common_pruning.py [hunk 4]
```diff
@@ -8,7 +8,16 @@
...
+    i = 0
+    for row in subset_tensor:
+        while i < len(superset_tensor):
+            if not torch.equal(row, superset_tensor[i]):
+                i += 1
+            else:
+                break
+        else:
+            return False
+    return True
```
Comment (WongJohnson) [inline torch/testing/_internal/common_pruning.py:20]: Some comments regarding assumptions (i.e. subset and order preserved) would help future use of this function, since it no longer matches the function name.

**Review by WongJohnson — COMMENTED** [review]
Thanks for incorporating some of my suggestions.

File: torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+                # handle weights in 4 groups
+                split_size = len(mask) // 4
+                masks = torch.split(mask, split_size)
+                saliencies = torch.split(saliency, split_size)
+
+                for keep_mask, sal in zip(masks, saliencies):
+                    # mask smallest k values to be removed
+                    k = int(len(keep_mask) * kwargs["sparsity_level"])
+                    prune = sal.topk(k, largest=False, sorted=False).indices
```
Comment (WongJohnson) [inline torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py:35]: Perhaps pull this logic out (minus the 4-split) and share between `SaliencyPruner` and `LSTMSaliencyPruner`

File: test/ao/sparsity/test_structured_sparsifier.py (hunk 3 above)
Comment (jcaip) [reply test/ao/sparsity/test_structured_sparsifier.py:98]: I think actually pruning 1d tensors is something that should not work - since we do rowise pruning, it doesn't really make sense to prune a 1d tensor. When we prune the bias, we're actually just taking the mask from the weight and applying it to the bias tensor, not computing a mask for the bias itself. This should probably throw a clearer error message earlier though.

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 6]
```diff
@@ -99,6 +103,10 @@ def _get_default_structured_pruning_patterns() -> Dict[
         # conv2d -> conv2d
         (nn.Conv2d, "output"): prune_conv2d,
         (nn.Conv2d, nn.Conv2d): prune_conv2d_conv2d,
+        # lstm -> getitem -> linear
+        (nn.LSTM, getitem, nn.Linear): prune_lstm_linear,
```
Comment (jerryzh168) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:107]: I remember the output of LSTM is `output, (h_n, c_n)`, so there will be multiple getitem node after the call to forward, it won't affect the current use case much probably since getitem is followed by some other ops, but this is not very robust I think

File: torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+                # select weights based on magnitude
+                if weights.dim() <= 1:
+                    raise Exception("Structured pruning can only be applied to a 2+dim weight tensor!")
+                # take norm over all but first dim
+                dims = tuple(range(1, weights.dim()))
+                saliency = weights.norm(dim=dims, p=1)
+
+                # handle weights in 4 groups
+                split_size = len(mask) // 4
```
Comment (jerryzh168) [inline torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py:26]: why do we handle weights in 4 groups?

Comment (WongJohnson) [reply torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py:26]: There are four sets of weights in each tensor. It may be worth verifying whether they should be pruned completely independently.
https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#:~:text=of%20shape%20(-,4*hidden_size,-%2C%20input_size)%20for
(search for "4*hidden_size" if the link doesn't jump for you)

File: torch/testing/_internal/common_pruning.py [hunk 8]
```diff
@@ -7,6 +7,22 @@
 from torch import nn
 
 
+def rows_are_subset(subset_tensor, superset_tensor) -> bool:
+    """
+    Checks to see if all rows in subset tensor are present in the superset tensor
```
Comment (WongJohnson) [inline torch/testing/_internal/common_pruning.py:12]: Preserved ordering of the subset tensor is the second assumption I was referring to - that is the basis of the speed-up compared to the original implementation.

File: torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py (hunk 7 above)
Comment (jerryzh168) [reply torch/ao/pruning/_experimental/pruner/lstm_saliency_pruner.py:26]: I see, thanks, would be helpful to add a comment I think

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py (hunk 6 above)
Comment (jcaip) [reply torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:107]: Yeah, this is very brittle - I tried to match `getitem(0)` instead of `getitem` only but couldn't figure out how to do that, if you have know how?

I can rename this to `prune_lstm_output_linear` I think to make it a bit more clear it's only for lstm -> output -> linear and not lstm -> h_n / c_n -> linear. The pruning function implementation is only correct lstm-> output -> linear anyways, it applies the wrong mask for h_n / c_n.

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42199001).

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 9]
```diff
@@ -83,14 +87,14 @@ def _get_supported_activation_modules():
...
-    Tuple[Union[Type[nn.Module], Callable[[torch.Tensor], torch.Tensor], str], ...],
+    Tuple[Union[Type[nn.Module], Callable, MatchAllNode, str], ...],
     Callable[..., None],
 ]:
     """
     Returns the patterns for conv2d / linear conversion for each element in the activation functions/modules defined above.
     """
     patterns: Dict[
-        Tuple[Union[Type[nn.Module], Callable[[torch.Tensor], torch.Tensor], str], ...],
+        Tuple[Union[Type[nn.Module], Callable, MatchAllNode, str], ...],
```
Comment (jerryzh168) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:97]: is this change needed? didn't see MatchAllNode being used

**Review by jerryzh168 — APPROVED** [review]
LGTM, but maybe @WongJohnson should take a final look since I'm not very familiar with  details
